src/server/engine_shard_set.cc

```diff
@@ -39,8 +49,260 @@ using absl::GetFlag;
+
+auto ReallocateValueImpl(PrimeValue& current_value) -> bool {
+  // we know that this is a string, we checked that before
+  auto val = current_value.ToString();
+  // we can now reset the old object as we are going to create a new one
+  current_value.Reset();
+  // This should allocate a new value
+  current_value.SetString(val);
+  return current_value.MallocUsed() > 0;
+}
+
+auto ReallocateValue(PrimeIterator it, void* from, const MemoryDefragParams& params, DbIndex index)
+    -> bool {
+  // allocate a new memory for the value, and then free the memory that used for the old one
+  CHECK(it->second.ObjType() == OBJ_STRING);  // only supporting for string right now!
+  if (it->second.IsExternal()) {
+    VLOG(1) << std::hex << from << ": the object that we are dealing with is external for type "
+            << std::dec << OpTypeToString(params.object_type)
+            << " or was not allcoated on the heap";
+    return false;
+  }
+
+  params.slice->PreUpdate(index, it);
+  auto key = it->first.ToString();
+  auto state = ReallocateValueImpl(it->second);
+  params.slice->PostUpdate(index, it, key);
+  return state;
+}
+
+auto MemoryReallocate(PrimeIterator it, const MemoryDefragParams& params, DbIndex index)
+    -> MemDefragState {
+  if (!IsValid(it)) {
+    return MemDefragState::REQUIRED;
+  }
+  if (MatchType(it->second.ObjType(), params.object_type)) {
+    auto ptr = it->second.RObjPtr();
+    return ReallocateValue(it, ptr, params, index) ? MemDefragState::REQUIRED
+                                                   : MemDefragState::NOT_REQUIRED;
+  }
+  return MemDefragState::NOT_REQUIRED;
+}
+
+auto DefragMemory(const MemoryDefragParams& params, DbIndex index)
+    -> std::tuple<MemDefragState, uint64_t> {
+  constexpr size_t CURSOR_COUNT_RELATION = 50;
+
+  auto shard_id = params.slice->shard_id();
+  DCHECK(params.slice->IsDbValid(index));
+
+  unsigned cnt = 0;
+  std::size_t max = params.slice->DbSize(index) / CURSOR_COUNT_RELATION;
+  VLOG(1) << "PrimeTable shared id/index " << params.slice->shard_id() << "/" << index << " has "
+          << params.slice->DbSize(index) << " elements in it";
+
+  PrimeTable::Cursor cur = params.cursor;
+  auto [prime_table, expire_table] = params.slice->GetTables(index);
+  MemDefragState state = max > 0 ? MemDefragState::REQUIRED : MemDefragState::NOT_REQUIRED;
+
+  while (cnt < max && state != MemDefragState::NOT_REQUIRED) {
+    cur = prime_table->Traverse(
+        cur, [&](PrimeIterator it) { state = MemoryReallocate(it, params, index); });
+    if (!cur) {
+      state = MemDefragState::NOT_REQUIRED;
+    }
+  }
+
+  if (cnt > 0) {
+    VLOG(1) << shard_id << ": ran the memory re-ordering task for " << cnt << " times - cursor "
+            << cur.value();
+  }
+
+  return {state, cur.value()};
+}
+
+auto roundOff(double n) -> double {
+  double d = n * 100.0;
+  int i = d + 0.5;
+  d = (float)i / 100.0;
+  return d;
+}
+
+auto Mem2String(uint64_t mem) -> std::string {
```

Comment:
we already have a function that prints it. look for `HumanReadableNumBytes`.

Reply:
removed